feat: Grafana dashboards for visualizing the sensor analytics

sanyamjain300 opened the pull request:
## PR Checklist
<!-- Please check if your PR fulfills the following requirements: -->

- [ ] Added label to the Pull Request for easier discoverability and search
- [x] Commit Message meets guidelines as indicated in the URL https://github.com/intel-retail/automated-self-checkout/blob/main/CONTRIBUTING.md
- [x] Every commit is a single defect fix and does not mix feature addition or changes
- [ ] Unit Tests have been added for new changes
- [x] Updated Documentation as relevant to the changes
- [x] All commented code has been removed
- [x] If you've added a dependency, you've ensured license is compatible with repository license and clearly outlined the added dependency.
- [ ] PR change contains code related to security
- [ ] PR introduces changes that breaks compatibility with other modules (If YES, please provide details below)

## What are you changing?
Create a Flask application which provides a grafana dashboard using API links and also automatically update the sensor data we used.

## Issue this PR will close

close: #542 

## Anything the reviewer should know when reviewing this PR?
It is a flask application not integrated.

## Test Instructions if applicable
<!-- How can the reviewers test your change? -->

## If the there are associated PRs in other repositories, please link them here (i.e. intel-retail/automated-self-checkout )

NeethuES-intel commented:
@sanyamjain300 Can you provide the steps to test this app ?
Also can you fix the security issues flagged by the github security bot/

sanyamjain300 commented:
Yeah sure, we will make the required changes and update the code.

ksm007 commented:
Hey I have changed the necessary files and added readme file in the newly created folder. can you check if everything is fine or is there any changes needs to be done

NeethuES-intel commented on src/retail-data-visualization/Dockerfile.retail-data-visualization:
```diff
@@ -0,0 +1,25 @@
+#
+# Copyright (C) 2024 Intel Corporation.
```
```suggestion
# Copyright (C) 2025 Intel Corporation.
```

ksm007 replied:
Hey can you check if plugin Infinity is installed in Administration -> Plugin and Data
I am installing it programmatically and I have tested it

ksm007 replied:
I have added make command to install the plugin manually to manually install plugin 
make install-infinity

NeethuES-intel commented on src/retail-data-visualization/dummy_data_load.py:
```diff
@@ -0,0 +1,111 @@
+from flask import Flask, jsonify
```
this file needs the copywrite -
```
#
# Copyright (C) 2025 Intel Corporation.
#
# SPDX-License-Identifier: Apache-2.0
#
```

ksm007 replied:
Added the changes

NeethuES-intel commented on src/retail-data-visualization/grafana/dashboards/dashboard.yaml:
this file is empty do we need this ?

ksm007 replied:
Removed this

NeethuES-intel commented on src/retail-data-visualization/grafana/dashboard.yaml:
```diff
@@ -0,0 +1,12 @@
+    updateIntervalSeconds: 10
+    allowUiUpdates: true
+    options:
+      path: /var/lib/grafana/dashboards
```
may be datasource needs to be configured here when dashboard gets initialized. I am currently getting error - "Datasource infinity was not found"

NeethuES-intel commented on src/retail-data-visualization/readme.md:
````diff
@@ -0,0 +1,54 @@
+Navigate to retail-data-visualization and run the command
+
+```
+docker-compose -f docker-compose.datavis.yml build --no-cache
````
can you create a Makefile under retail-data-visualization and add these as simple make commands eg - https://github.com/intel-retail/automated-self-checkout/blob/48f15398a188c00d6fc921f2d7199e5328141d7f/Makefile#L35

ksm007 replied:
added Makefile to build
build -
**make build**
run-
**make run**

NeethuES-intel commented on src/retail-data-visualization/readme.md:
````diff
@@ -0,0 +1,54 @@
+### 1. Run the application
+
+```
+docker compose -f docker-compose.datavis.yml up -d
````
same comment as above, make it simple command inside Makefile

NeethuES-intel reviewed (changes requested):
@sanyamjain300 @ksm007 @RushabhGajab2605 I left comments on the code. 
While running the code I am getting the following error - empty dashboard with error "Datasource infinity was not found"
Can you initialize the dashboard directly with the datasource when it is opened for the first time ?
![image](https://github.com/user-attachments/assets/8f51a9c8-e535-41de-afc9-2374c89e886c)

NeethuES-intel commented on src/retail-data-visualization/Makefile:
```diff
@@ -0,0 +1,11 @@
+build:
```
```suggestion
build: install-infinity
```
this will automatically line 4, user doesn't have to do any extra step then

NeethuES-intel commented on src/retail-data-visualization/readme.md:
```diff
@@ -0,0 +1,54 @@
+## Features
```
Can you update the readme.md with Makefile changes, build & run commands that are changed. Also rename readme.md as README.md

NeethuES-intel reviewed (changes requested):
I got error with make install-infinity
![image](https://github.com/user-attachments/assets/7d923e4e-e976-4b02-b238-a9f00463f629)

Also I do see the plugin installed under Administration. Am I missing some step to run this correctly ?
![image](https://github.com/user-attachments/assets/6c45b8dd-fd46-464a-b0aa-c698cc268c58)

antoniomtz commented:
![image](https://github.com/user-attachments/assets/deabfe66-9494-4817-8ca8-790dcd06d891)
I was able to run it. For each graph in the dashboard, I need to click edit, wait 2-3 seconds and it will refresh the graph. At the beginning, it will show "no data". @ksm007

ksm007 commented:
Ok... I just noticed that issue for first run this is occurring, I will fix it and push the fix soon
Thanks @antoniomtz

RushabhGajab2605 commented:
We have fixed the issue. Can you pull and check @NeethuES-intel @antoniomtz ?

ksm007 commented:
I guess this is browser issue...

NeethuES-intel reviewed (commented):
@sanyamjain300 @ksm007 @RushabhGajab2605 Awesome job getting this feature ready for merge. Everything worked for me on my machine. PR is 90% complete. Please work on the pending review comments so that we can get this feature merged in.

NeethuES-intel commented:
@sanyamjain300 @ksm007 @RushabhGajab2605 Also please note, branch is out-of-date with main. Please merge main into your branch.

NeethuES-intel commented:
@sanyamjain300 @ksm007 @RushabhGajab2605 Please merge in main & make the branch up to date. After addressing the remaining  PR comments, your PR is ready for merge.

sanyamjain300 commented:
> @sanyamjain300 @ksm007 @RushabhGajab2605 Please merge in main & make the branch up to date. After addressing the remaining PR comments, your PR is ready for merge.

Done

NeethuES-intel commented:
> > @sanyamjain300 @ksm007 @RushabhGajab2605 Please merge in main & make the branch up to date. After addressing the remaining PR comments, your PR is ready for merge.
> 
> Done

Thank you @sanyamjain300. Can you fix the security issues flagged by reviewdog ?

sanyamjain300 commented:
> Thank you @sanyamjain300. Can you fix the security issues flagged by reviewdog ?

Sorry, I am not sure about these flagged issues.

NeethuES-intel commented on src/retail-data-visualization/dummy_data_load.py:
```diff
@@ -0,0 +1,111 @@
+    thread.start()
+
+if __name__ == '__main__':
+    app.run(host='0.0.0.0', port=8000)
```
bandit is flagging following error - B104:MEDIUM:Possible binding to all interfaces.
https://bandit.readthedocs.io/en/latest/plugins/b104_hardcoded_bind_all_interfaces.html
How about using 127.0.0.1 or localhost ?

sanyamjain300 replied:
Yeah, we can use 127.0.0.1 instead.

NeethuES-intel commented on src/retail-data-visualization/dummy_data_load.py:
```diff
@@ -0,0 +1,111 @@
+    for i in range(1, 21):
+        barcode_data.append({
+            "sensor_id": f"barcode_{i}",
+            "item_code": f"code_{random.randint(1000, 9999)}",
```
we can ignore this, since random is used here for generating dummy sensor data.

ksm007 replied:
remove for all random generation right?

NeethuES-intel replied:
> remove for all random generation right?

you can ignore this warning, since we are not using this function for security. Another option, just add 20 random hard coded values, instead of using random() to completely remove this error.

NeethuES-intel replied:
> remove for all random generation right?

@ksm007 check this PR, they are using numpy.random.uniform to simulate data. 
https://github.com/intel-retail/automated-self-checkout/blob/6b526762f8847ada6e2475a9522c724dced01790/src/common-service/lidar_app.py#L44

NeethuES-intel commented:
> > Thank you @sanyamjain300. Can you fix the security issues flagged by reviewdog ?
> 
> Sorry, I am not sure about these flagged issues.

@sanyamjain300 @ksm007 @RushabhGajab2605 I added comments, can you work on them ? You can ignore the low severity issue.

antoniomtz reviewed (approved):
LGTM. I was able to run and see grafana

NeethuES-intel commented on src/retail-data-visualization/readme.md:
```diff
@@ -0,0 +1,59 @@
+## Features
```
@sanyamjain300 @ksm007 @RushabhGajab2605 One more request, can you change this filename to README.md ? following the same naming convention across repos.

ejlee3 commented on src/retail-data-visualization/readme.md:
```diff
@@ -0,0 +1,59 @@
+
+### 1. Build the application
+
+Navigate to retail-data-visualization and run the command
```
put the src folder here

```suggestion
Navigate to `src/retail-data-visualization` and run the command
```

ejlee3 reviewed (changes requested):
one small change in docs. Once that's done, nice work - it worked on my system!

NeethuES-intel commented:
@sanyamjain300 @ksm007 @RushabhGajab2605 Our sprint is ending next week. We would love to have your PRs merged in before that. Can you help addressing the PR comments ?

sanyamjain300 commented:
@NeethuES-intel really sorry for the delay caused. We will be resolving these issues as soon as possible.

NeethuES-intel commented on src/retail-data-visualization/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+Flask==2.2.5
+Flask-Cors==4.0.2
+numpy==1.26.4
+gunicorn==20.1.0
```
@sanyamjain300 Good job resolving reviewdog issues. However, I see this high risk CVE flagged by security-bot. Can you please fix this. Your PR is ready for merge after resolving this issue

ksm007 commented:
@NeethuES-intel  I have done the necessary changes and pushed the code, can you verify this once, Thank you

NeethuES-intel commented:
> @NeethuES-intel I have done the necessary changes and pushed the code, can you verify this once, Thank you

@sanyamjain300 @ksm007 @RushabhGajab2605 Please merge the latest changes, branch is out-of-date

NeethuES-intel reviewed (approved):
Good job closing on the review comments.

ejlee3 reviewed (approved):
Nice work!

NeethuES-intel commented:
Good job @sanyamjain300 @ksm007 @RushabhGajab2605 for working and closing on this issue.

sanyamjain300 commented:
Thank you very much @NeethuES-intel @ejlee3, it was our pleasure to work on this issue.
